Allow for special constant to be non-conformant with object's data type

al-niessner opened the pull request:
## 🗒️ Summary
Added unit test then fix. Basically check on conversion if value matches a data type non-conforming special constant before logging error.

## ⚙️ Test Data and/or Report
See automated unit/regression testing below

## ♻️ Related Issues
Closes #837

Commit:
add unit tests

Commit:
fix

nutjob4life commented on src/main/java/gov/nasa/pds/tools/validate/SpecialConstantChecker.java:
```diff
@@ -0,0 +1,172 @@
+          comparison = Double.valueOf(value.doubleValue())
+              .compareTo(SpecialConstantBitPatternTransforms.asBigDecimal(constants.getValidMinimum()).doubleValue());
+        } else {
+          Long con = SpecialConstantBitPatternTransforms.asBigInt(constants.getValidMinimum()).longValue();
```
"Long con" … reminds me of a past life 🤭

nutjob4life reviewed (approved):
👍

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/SpecialConstantChecker.java:
```diff
@@ -0,0 +1,172 @@
+   * @param constants
+   * @return
+   */
+  public static boolean isNonConformantSpecialConstant(String value, SpecialConstants constants) {
```
@al-niessner so per the [PDS4 IM Spec](https://pds.nasa.gov/datastandards/documents/im/v1/index_1L00.html), these are the special constants that do not require matching data types:
* error_constant - https://pds.nasa.gov/datastandards/documents/im/v1/index_1L00.html#attribute_pds_special_constants_pds_error_constant
* invalid_constant - https://pds.nasa.gov/datastandards/documents/im/v1/index_1L00.html#attribute_pds_special_constants_pds_invalid_constant
* missing_constant - https://pds.nasa.gov/datastandards/documents/im/v1/index_1L00.html#attribute_pds_special_constants_pds_missing_constant
* not_applicable_constant - https://pds.nasa.gov/datastandards/documents/im/v1/index_1L00.html#attribute_pds_special_constants_pds_not_applicable_constant
* saturated_constant - https://pds.nasa.gov/datastandards/documents/im/v1/index_1L00.html#attribute_pds_special_constants_pds_saturated_constant
* unknown_constant - https://pds.nasa.gov/datastandards/documents/im/v1/index_1L00.html#attribute_pds_special_constants_pds_unknown_constant

The rest all state that they must match the data type of the object and should throw an error if they do not match.

al-niessner replied:
done

al-niessner committed:
Update SpecialConstantChecker.java

Comment out the ones that we are not allowed today but will have to do tomororw.

jordanpadams committed:
Merge branch 'main' into issue_837

Commit:
fix what looks like bad merge
